### PR TITLE
Extract the ListenForMessages method to his own class

### PR DESCRIPTION
**Pull request by aboimpinto**
This method does not interact with Server and can be extracted and tested separately as a module.

This PR is not complete. It's important to all check the change and propose changes.

### CODE REVIEW

**Comment by shargon** [comment]
Should be great if we can raise new connection from this class, and this class could be injected inside GetPeers message response. We need to solve this

**Comment by aboimpinto** [comment]
I don't think if I follow you ... 
can you elaborate a little bit better?

**Comment by shargon** [comment]
How can i connect one new `IPeer` from `AddrMessageMessageHandler`?

**Comment by aboimpinto** [comment]
Tkx for the review ...

Right now, this functionality do not interact in any way with Server class. After been called in the Server for every message received from the Peer, this method just use the MessageHandler to process the message.
The other way the Server interact with the connections is when to cancel it.

Because all of this, this is a excellent case where we can extract this method to is own class that will provide cleaner code on the Server class.

Answering to your specific questions
1º When we run the test we need to setup this methods that will  receive message from the peer and /or we need to establish that all this tasks are started before stop the server (some tests on the stop).
This situations are caused by the nature of this code that can be easy put out of the class due the fact there is no interaction with any other method of the Server class.
With this fix, we just need to make sure that we mock the IPeerMessageListener and make sure the Start method is called.
Remove the instability of the tests.

2º Right now this is not implemented but to implement this, as soon we have an error on the handshake we can raise an event that will be register and catch in the Server class and take the right actions, like disconnect the corrected peer.

With this change we are creating smaller classes with specific functions and the Server will be the maestro organizing when to call this or that and reacting with the event that receive.

**Comment by osmirnov** [comment]
Ok, thanks for answering my questions and this is quite close to my assumptions:
1. Server accepts peers but has no idea what to do with them.
2. PeerMessageListener makes no sense without Server and Server being in running state.
3. PeerMessageListener does not track connected peers. It means we cannot send a message to specific peer, broadcasting is impossible, a peer cannot be disconnected, etc.
4. We keep increasing server dependencies introducing classes for special cases as ServerContext and PeerMessageListener.
5. To eliminate instability in tests we mock it instead of treating that as a real business scenario ready to be tested.
6.  As we need to extend special case classes we extend Server contract. For example, the suggested solution about the introduction of events for misbehavior detection creates on more dependency (1st: constructor injection of IPeerMessageListener interface, 2nd: IPeerMessageListener.OnPeerFailed proposed event to fire if handshake or any other procotol was violated).

To sum up, I agree currently we do not interact with Server in message exchange but it can easily change in near future and keep tearing that apart per specific task increasing complexity as well as support cost.
